File: Modules/WillStrohl.Injection/EditInjections.ascx.vb
'
' Will Strohl (will.strohl@gmail.com)
' http://www.willstrohl.com
'
'Copyright (c) 2009-2015, Will Strohl
'All rights reserved.
'
'Redistribution and use in source and binary forms, with or without modification, are 
'permitted provided that the following conditions are met:
'
'Redistributions of source code must retain the above copyright notice, this list of 
'conditions and the following disclaimer.
'
'Redistributions in binary form must reproduce the above copyright notice, this list 
'of conditions and the following disclaimer in the documentation and/or other 
'materials provided with the distribution.
'
'Neither the name of Will Strohl, Content Injection, nor the names of its contributors may be 
'used to endorse or promote products derived from this software without specific prior 
'written permission.
'
'THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY 
'EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES 
'OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT 
'SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
'INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED 
'TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR 
'BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
'CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
'ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH 
'DAMAGE.
'

Imports System.Text.RegularExpressions
Imports System.Web.UI.WebControls
Imports DotNetNuke
Imports DotNetNuke.Services.Exceptions
Imports DotNetNuke.Security.PortalSecurity
Imports DotNetNuke.UI.Skins.Skin

Namespace WillStrohl.Modules.Injection

    Partial Public MustInherit Class EditInjections
        Inherits WNSPortalModuleBase

#Region " Private Members "

        Private p_Injections As InjectionInfoCollection = Nothing

        Private Const c_Command_Edit As String = "Edit"
        Private Const c_Command_MoveUp As String = "MoveUp"
        Private Const c_Command_MoveDown As String = "MoveDown"
        Private Const c_Command_Delete As String = "Delete"
        Private Const c_Command_Insert As String = "Insert"

        Private Const c_True As String = "True"

        Private p_SearchParam As Integer = Null.NullInteger

        Private p_EnabledImage As String = String.Empty
        Private p_DisabledImage As String = String.Empty
        Private p_EnabledAltText As String = String.Empty
        Private p_DisabledAltText As String = String.Empty

#End Region

#Region " Private Properties "

        Private ReadOnly Property Injections() As InjectionInfoCollection
            Get
                If Me.p_Injections Is Nothing Then
                    Dim ctlModule As New InjectionController
                    Me.p_Injections = ctlModule.GetInjectionContents(Me.ModuleId)
                End If
                Return Me.p_Injections
            End Get
        End Property

        Private ReadOnly Property EnabledImage() As String
            Get
                If Not String.IsNullOrEmpty(p_EnabledImage) Then
                    Return p_EnabledImage
                End If

                p_EnabledImage = String.Concat(Common.Globals.ApplicationPath, Entities.Icons.IconController.IconURL("Checked", "16x16"))

                Return p_EnabledImage
            End Get
        End Property

        Private ReadOnly Property DisabledImage() As String
            Get
                If Not String.IsNullOrEmpty(p_DisabledImage) Then
                    Return p_DisabledImage
                End If

                p_DisabledImage = String.Concat(Common.Globals.ApplicationPath, Entities.Icons.IconController.IconURL("Unchecked", "16x16"))

                Return p_DisabledImage
            End Get
        End Property

        Private ReadOnly Property EnabledAltText() As String
            Get
                If String.IsNullOrEmpty(Me.p_EnabledAltText) Then
                    Me.p_EnabledAltText = Me.GetLocalizedString("Enabled.Alt")
                End If

                Return Me.p_EnabledAltText
            End Get
        End Property

        Private ReadOnly Property DisabledAltText() As String
            Get
                If String.IsNullOrEmpty(Me.p_DisabledAltText) Then
                    Me.p_DisabledAltText = Me.GetLocalizedString("Disabled.Alt")
                End If

                Return Me.p_DisabledAltText
            End Get
        End Property
        
#End Region
        
#Region " Event Handlers "

        Private Sub Page_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Load
            Try
                If Not Me.Page.IsPostBack Then
                    Me.pnlAddNew.Visible = False
                    Me.pnlManage.Visible = True

                    Me.BindData()
                End If
            Catch exc As Exception ' Module failed to load
                ProcessModuleLoadException(Me, exc, Me.IsEditable)
            End Try
        End Sub

        Private Sub lnkAddNewInjection_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles lnkAddNewInjection.Click, cmdCancel.Click

            Me.ClearForm()
            Me.TogglePanels()

            Me.cmdDelete.Visible = Not String.IsNullOrEmpty(Me.hidInjectionId.Value)

        End Sub

        Private Sub cmdReturn_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles cmdReturn.Click
            Response.Redirect(NavigateURL)
        End Sub

        Private Sub cmdAdd_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles cmdAdd.Click

            If Me.Page.IsValid AndAlso Not (String.IsNullOrEmpty(Me.txtName.Text) Or String.IsNullOrEmpty(Me.txtContent.Text)) Then
                Me.SaveInjection()
                Me.ClearForm()
                Me.TogglePanels()
                Me.BindData()
            End If

        End Sub

        Private Sub cvName_ServerValidate(ByVal source As Object, ByVal args As System.Web.UI.WebControls.ServerValidateEventArgs) Handles cvName.ServerValidate
            Dim ctlModule As New InjectionController
            args.IsValid = (Not ctlModule.DoesInjectionNameExist(Me.txtName.Text, Me.ModuleId))
        End Sub

        Protected Sub dlInjection_ItemCommand(ByVal source As Object, ByVal e As System.Web.UI.WebControls.DataListCommandEventArgs) Handles dlInjection.ItemCommand
            Select Case e.CommandName
                Case c_Command_MoveUp
                    Me.SwapOrder(Convert.ToInt32(e.CommandArgument), c_Command_MoveUp)
                    Me.BindData()
                Case c_Command_MoveDown
                    Me.SwapOrder(Convert.ToInt32(e.CommandArgument), c_Command_MoveDown)
                    Me.BindData()
                Case c_Command_Edit
                    Me.BindForm(Convert.ToInt32(e.CommandArgument))
                    Me.TogglePanels()
                Case c_Command_Insert
                    Me.ClearForm()
                    Me.TogglePanels()
                Case c_Command_Delete
                    Dim ctlModule As New InjectionController
                    ctlModule.DeleteInjectionContent(Convert.ToInt32(e.CommandArgument))
                    Me.BindData()
                Case Else
                    Exit Sub
            End Select
        End Sub

        Private Sub cmdDelete_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles cmdDelete.Click

            If Regex.IsMatch(Me.hidInjectionId.Value, "^\d+$") Then
                Dim ctlModule As New InjectionController
                ctlModule.DeleteInjectionContent(Integer.Parse(Me.hidInjectionId.Value, Globalization.NumberStyles.Integer))
            End If

            Me.ClearForm()
            Me.TogglePanels()
            Me.BindData()

        End Sub

#End Region

#Region " Private Helper Functions "

        Private Sub BindData()

            Me.LocalizeModule()

            ' bind data to controls

            If Me.Injections.Count > 0 Then
                With Me.dlInjection
                    .DataSource = Injections
                    .DataBind()
                End With
                Me.dlInjection.Visible = True
                Me.lblNoRecords.Visible = False
            Else
                Me.dlInjection.Visible = False
                Me.lblNoRecords.Visible = True
            End If

            Me.txtName.Attributes.Add("onfocus", String.Concat("if (this.value == '", Me.GetLocalizedString("txtName.Text"), "') { this.value = ''; }"))

            If Me.radInject.Items.Count = 0 Then
                Me.radInject.Items.Add(New ListItem(Me.GetLocalizedString("radInject.0.Text")))
                Me.radInject.Items.Add(New ListItem(Me.GetLocalizedString("radInject.1.Text")))
            End If

        End Sub

        Private Sub LocalizeModule()

            Me.txtName.Text = Me.GetLocalizedString("txtName.Text")
            Me.lnkAddNewInjection.Text = Me.GetLocalizedString("lnkAdd.Text")
            Me.cmdAdd.Text = Me.GetLocalizedString("cmdAdd.Text")
            Me.cmdDelete.Text = Me.GetLocalizedString("cmdDelete.Text")
            Me.rfvName.ErrorMessage = Me.GetLocalizedString("rfvName.ErrorMessage")
            Me.rfvName.InitialValue = Me.GetLocalizedString("txtName.Text")
            Me.rfvContent.ErrorMessage = Me.GetLocalizedString("rfvContent.ErrorMessage")
            Me.cmdCancel.Text = Me.GetLocalizedString("cmdCancel.Text")
            Me.cmdReturn.Text = Me.GetLocalizedString("cmdReturn.Text")
            Me.cvName.ErrorMessage = Me.GetLocalizedString("cvName.ErrorMessage")

        End Sub

        Private Sub ClearForm()
            Me.hidInjectionId.Value = String.Empty
            Me.txtName.Text = Me.GetLocalizedString("txtName.Text")
            Me.txtContent.Text = String.Empty
            Me.chkEnabled.Checked = True
            Me.radInject.Items.FindByText(Me.GetLocalizedString("radInject.0.Text")).Selected = True
            Me.cvName.Enabled = True
        End Sub

        Private Sub BindForm(ByVal ItemId As Integer)

            Dim ctlModule As New InjectionController
            Dim objInfo As New InjectionInfo
            objInfo = ctlModule.GetInjectionContent(ItemId)

            Me.txtName.Text = objInfo.InjectName
            Me.txtContent.Text = Server.HtmlDecode(objInfo.InjectContent)
            Me.radInject.ClearSelection()
            If objInfo.InjectTop Then
                Me.radInject.Items.FindByText(Me.GetLocalizedString("radInject.0.Text")).Selected = True
            Else
                Me.radInject.Items.FindByText(Me.GetLocalizedString("radInject.1.Text")).Selected = True
            End If
            Me.chkEnabled.Checked = objInfo.IsEnabled
            Me.hidInjectionId.Value = objInfo.InjectionId.ToString
            Me.cvName.Enabled = False

            Me.cmdDelete.Visible = Not String.IsNullOrEmpty(Me.hidInjectionId.Value)

        End Sub

        Private Sub SwapOrder(ByVal ItemId As Integer, ByVal UpDown As String)
            
            ' set the global id to match the one we're looking for
            Me.p_SearchParam = ItemId

            ' change the order
            Dim ctlModule As New InjectionController
            ctlModule.ChangeOrder(ItemId, UpDown)

        End Sub

        Private Function FindInjectionById(ByVal item As InjectionInfo) As Boolean
            Return item.InjectionId = Me.p_SearchParam
        End Function

        Private Sub TogglePanels()
            Me.pnlAddNew.Visible = (Not Me.pnlAddNew.Visible)
            Me.pnlManage.Visible = (Not Me.pnlManage.Visible)
        End Sub

        Private Sub SaveInjection()
            Try
                Dim sec As New Security.PortalSecurity
                Dim ctlModule As New InjectionController
                Dim objInj As New InjectionInfo

                If Not String.IsNullOrEmpty(Me.hidInjectionId.Value) Then
                    objInj = ctlModule.GetInjectionContent(Integer.Parse(Me.hidInjectionId.Value))
                    With objInj
                        .InjectContent = Server.HtmlEncode(Me.txtContent.Text)
                        .InjectName = sec.InputFilter(Me.txtName.Text, FilterFlag.NoMarkup)
                        .InjectTop = Me.radInject.Items.FindByText(Me.GetLocalizedString("radInject.0.Text")).Selected
                        .IsEnabled = Me.chkEnabled.Checked
                        .ModuleId = Me.ModuleId
                    End With
                    ctlModule.UpdateInjectionContent(objInj)
                Else
                    With objInj
                        .InjectContent = Server.HtmlEncode(Me.txtContent.Text)
                        .InjectName = sec.InputFilter(Me.txtName.Text, FilterFlag.NoMarkup)
                        .InjectTop = Me.radInject.Items.FindByText(Me.GetLocalizedString("radInject.0.Text")).Selected
                        .IsEnabled = Me.chkEnabled.Checked
                        .ModuleId = Me.ModuleId
                        .OrderShown = ctlModule.GetNextOrderNumber(Me.ModuleId)
                    End With
                    ctlModule.AddInjectionContent(objInj)
                End If
            Catch ex As Exception
                HandleException(ex)
            End Try
        End Sub

        Public Function CommandUpVisible(ByVal InjectionId As String) As String

            Dim ctlModule As New InjectionController
            Dim oInject As InjectionInfo = ctlModule.GetInjectionContent(Integer.Parse(InjectionId, Globalization.NumberStyles.Integer))
            
            Return (Not oInject.OrderShown = 1).ToString

        End Function

        Public Function CommandDownVisible(ByVal InjectionId As String) As String

            Dim ctlModule As New InjectionController
            Dim oInject As InjectionInfo = ctlModule.GetInjectionContent(Integer.Parse(InjectionId, Globalization.NumberStyles.Integer))
            Dim collInject As InjectionInfoCollection = ctlModule.GetInjectionContents(Me.ModuleId)

            Return (Not oInject.OrderShown = collInject.Count).ToString

        End Function

        Public Function GetEnabledImage(ByVal EnabledText As String) As String
            If String.Equals(EnabledText, c_True) Then
                Return EnabledImage
            Else
                Return DisabledImage
            End If
        End Function

        Public Function GetEnabledImageAltText(ByVal EnabledText As String) As String
            If String.Equals(EnabledText, c_True) Then
                Return EnabledAltText
            Else
                Return DisabledAltText
            End If
        End Function

        Private Sub HandleException(ByVal exc As Exception)
            LogException(exc)
            If Me.UserInfo.IsSuperUser Or Me.UserInfo.UserID = PortalSettings.AdministratorId Then
                AddModuleMessage(Me, _
                    String.Concat(exc.Message, "<br />", exc.StackTrace), _
                    Skins.Controls.ModuleMessage.ModuleMessageType.RedError)
            Else
                AddModuleMessage(Me, exc.Message, _
                    Skins.Controls.ModuleMessage.ModuleMessageType.RedError)
            End If
        End Sub

#End Region

    End Class

End Namespace

File: Modules/WillStrohl.Injection/EditInjections.ascx.cs
/*
Copyright © Upendo Ventures, LLC

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and 
associated documentation files (the "Software"), to deal in the Software without restriction, 
including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, 
and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial 
portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT 
NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES 
OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Web.UI.WebControls;
using DotNetNuke.Common.Utilities;
using DotNetNuke.Services.Exceptions;
using System;
using System.Globalization;
using DotNetNuke.Entities.Users;
using DotNetNuke.Security;
using WillStrohl.Modules.Injection.Entities;
using WillStrohl.Modules.Injection.Components;

namespace WillStrohl.Modules.Injection
{
    public partial class EditInjections : WNSPortalModuleBase
    {
        #region Private Members

        private InjectionInfoCollection p_Injections = null;
        private const string c_Command_Edit = "Edit";
        private const string c_Command_MoveUp = "MoveUp";
        private const string c_Command_MoveDown = "MoveDown";
        private const string c_Command_Delete = "Delete";

        private const string c_Command_Insert = "Insert";

        private const string c_True = "True";

        private int p_SearchParam = Null.NullInteger;
        private string p_EnabledImage = string.Empty;
        private string p_DisabledImage = string.Empty;
        private string p_EnabledAltText = string.Empty;

        private string p_DisabledAltText = string.Empty;

        #endregion

        #region Private Properties

        private InjectionInfoCollection Injections
        {
            get
            {
                if (p_Injections == null)
                {
                    InjectionController ctlModule = new InjectionController();
                    p_Injections = ctlModule.GetInjectionContents(ModuleId);
                }
                return p_Injections;
            }
        }

        private string EnabledImage
        {
            get
            {
                if (!string.IsNullOrEmpty(p_EnabledImage))
                {
                    return p_EnabledImage;
                }

                p_EnabledImage = string.Concat(DotNetNuke.Common.Globals.ApplicationPath, DotNetNuke.Entities.Icons.IconController.IconURL("Checked", "16x16"));

                return p_EnabledImage;
            }
        }

        private string DisabledImage
        {
            get
            {
                if (!string.IsNullOrEmpty(p_DisabledImage))
                {
                    return p_DisabledImage;
                }

                p_DisabledImage = string.Concat(DotNetNuke.Common.Globals.ApplicationPath, DotNetNuke.Entities.Icons.IconController.IconURL("Unchecked", "16x16"));

                return p_DisabledImage;
            }
        }

        private string EnabledAltText
        {
            get
            {
                if (string.IsNullOrEmpty(p_EnabledAltText))
                {
                    p_EnabledAltText = GetLocalizedString("Enabled.Alt");
                }

                return p_EnabledAltText;
            }
        }

        private string DisabledAltText
        {
            get
            {
                if (string.IsNullOrEmpty(p_DisabledAltText))
                {
                    p_DisabledAltText = GetLocalizedString("Disabled.Alt");
                }

                return p_DisabledAltText;
            }
        }

        #endregion

        #region Event Handlers

        protected void Page_Load(object sender, EventArgs e)
        {
            try
            {
                if (!Page.IsPostBack)
                {
                    pnlAddNew.Visible = false;
                    pnlManage.Visible = true;

                    BindData();
                }
                // Module failed to load
            }
            catch (Exception exc)
            {
                Exceptions.ProcessModuleLoadException(this, exc, IsEditable);
            }
        }

        protected void lnkAddNewInjection_Click(object sender, EventArgs e)
        {
            ClearForm();
            TogglePanels();

            cmdDelete.Visible = !string.IsNullOrEmpty(hidInjectionId.Value);
        }

        protected void cmdReturn_Click(object sender, EventArgs e)
        {
            Response.Redirect(DotNetNuke.Common.Globals.NavigateURL());
        }

        protected void cmdAdd_Click(object sender, EventArgs e)
        {
            if (Page.IsValid && !(string.IsNullOrEmpty(txtName.Text) | string.IsNullOrEmpty(txtContent.Text)))
            {
                SaveInjection();
                ClearForm();
                TogglePanels();
                BindData();
            }
        }

        protected void cvName_ServerValidate(object source, ServerValidateEventArgs args)
        {
            InjectionController ctlModule = new InjectionController();
            args.IsValid = (!ctlModule.DoesInjectionNameExist(txtName.Text, ModuleId));
        }

        protected void dlInjection_ItemCommand(object source, DataListCommandEventArgs e)
        {
            switch (e.CommandName)
            {
                case c_Command_MoveUp:
                    SwapOrder(Convert.ToInt32(e.CommandArgument), c_Command_MoveUp);
                    BindData();
                    break;
                case c_Command_MoveDown:
                    SwapOrder(Convert.ToInt32(e.CommandArgument), c_Command_MoveDown);
                    BindData();
                    break;
                case c_Command_Edit:
                    BindForm(Convert.ToInt32(e.CommandArgument));
                    ToggleType();
                    TogglePanels();
                    break;
                case c_Command_Insert:
                    ClearForm();
                    TogglePanels();
                    break;
                case c_Command_Delete:
                    InjectionController ctlModule = new InjectionController();
                    ctlModule.DeleteInjectionContent(Convert.ToInt32(e.CommandArgument));
                    BindData();
                    break;
                default:
                    return;
                    break;
            }
        }

        protected void cmdCancel_Click(object sender, EventArgs e)
        {
            ClearForm();
            TogglePanels();
            BindData();
        }

        protected void cmdDelete_Click(object sender, EventArgs e)
        {
            if (Regex.IsMatch(hidInjectionId.Value, "^\\d+$"))
            {
                InjectionController ctlModule = new InjectionController();
                ctlModule.DeleteInjectionContent(int.Parse(hidInjectionId.Value, System.Globalization.NumberStyles.Integer));
            }

            ClearForm();
            TogglePanels();
            BindData();

        }

        protected void radType_OnSelectedIndexChanged(object sender, EventArgs e)
        {
            ToggleType();
        }

        protected void cvContent_OnServerValidate(object source, ServerValidateEventArgs args)
        {
            if (radType.SelectedIndex == 0)
            {
                var isValidPathFormat = false;
                var isValidFilePath = false;
                var pathToTest = txtContent.Text.Trim();

                isValidPathFormat = (InjectionController.IsValidCssInjectionType(pathToTest) ||
                                     InjectionController.IsValidJavaScriptInjectionType(pathToTest));

                if (pathToTest.StartsWith("http"))
                {
                    // external file path
                    isValidFilePath = InjectionController.IsValidFilePath(pathToTest);
                }
                else
                {
                    // local file path
                    var fullFilePath = Server.MapPath(pathToTest);
                    isValidFilePath = File.Exists(fullFilePath);
                }

                args.IsValid = (isValidPathFormat && isValidFilePath);
            }
            else
            {
                args.IsValid = true;
            }
        }

        #endregion

        #region Private Helper Functions

        private void BindData()
        {
            LocalizeModule();

            if (Injections.Count > 0)
            {
                dlInjection.DataSource = Injections;
                dlInjection.DataBind();
                dlInjection.Visible = true;
                lblNoRecords.Visible = false;
            }
            else
            {
                dlInjection.Visible = false;
                lblNoRecords.Visible = true;
            }

            if (radType.Items.Count == 0)
            {
                radType.Items.Add(new ListItem(GetLocalizedString("radType.0.Text")));
                radType.Items.Add(new ListItem(GetLocalizedString("radType.1.Text")));
                radType.ClearSelection();
                // default injection type for NEW injections is 0 because that's the most common use case
                radType.SelectedIndex = 0;
            }

            if (ddlCrmProvider.Items.Count == 0)
            {
                ddlCrmProvider.Items.Add(new ListItem(GetLocalizedString("ddlCrmProvider.0.Text")));
                ddlCrmProvider.Items.Add(new ListItem(GetLocalizedString("ddlCrmProvider.1.Text")));
                ddlCrmProvider.Items.Add(new ListItem(GetLocalizedString("ddlCrmProvider.2.Text")));
                ddlCrmProvider.Items.Add(new ListItem(GetLocalizedString("ddlCrmProvider.3.Text")));
                ddlCrmProvider.ClearSelection();
                ddlCrmProvider.SelectedIndex = 0;
            }

            if (radInject.Items.Count == 0)
            {
                radInject.Items.Add(new ListItem(GetLocalizedString("radInject.0.Text")));
                radInject.Items.Add(new ListItem(GetLocalizedString("radInject.1.Text")));
                radInject.ClearSelection();
                radInject.SelectedIndex = 0;
            }

        }

        private void LocalizeModule()
        {
            txtName.Text = GetLocalizedString("txtName.Text");
            lnkAddNewInjection.Text = GetLocalizedString("lnkAdd.Text");
            cmdAdd.Text = GetLocalizedString("cmdAdd.Text");
            cmdDelete.Text = GetLocalizedString("cmdDelete.Text");
            rfvName.ErrorMessage = GetLocalizedString("rfvName.ErrorMessage");
            rfvName.InitialValue = GetLocalizedString("txtName.Text");
            rfvContent.ErrorMessage = GetLocalizedString("rfvContent.ErrorMessage");
            cmdCancel.Text = GetLocalizedString("cmdCancel.Text");
            cmdReturn.Text = GetLocalizedString("cmdReturn.Text");
            cvName.ErrorMessage = GetLocalizedString("cvName.ErrorMessage");
            cvContent.ErrorMessage = GetLocalizedString("cvContent.ErrorMessage");
            rvCrmPriority.ErrorMessage = GetLocalizedString("rvCrmPriority.ErrorMessage");
        }

        private void ClearForm()
        {
            hidInjectionId.Value = string.Empty;
            txtName.Text = GetLocalizedString("txtName.Text");
            txtContent.Text = string.Empty;
            chkEnabled.Checked = true;
            radType.ClearSelection();
            radType.Items.FindByText(GetLocalizedString("radType.0.Text")).Selected = true;
            radInject.ClearSelection();
            radInject.Items.FindByText(GetLocalizedString("radInject.0.Text")).Selected = true;
            cvName.Enabled = true;
            ddlCrmProvider.ClearSelection();
            ddlCrmProvider.SelectedIndex = 0;

            ToggleType();
        }

        private void BindForm(int ItemId)
        {
            var ctlModule = new InjectionController();
            var injection = new InjectionInfo();

            injection = ctlModule.GetInjectionContent(ItemId);

            txtName.Text = injection.InjectName;
            txtContent.Text = Server.HtmlDecode(injection.InjectContent);
            radInject.ClearSelection();

            if (injection.InjectTop)
            {
                radInject.Items.FindByText(GetLocalizedString("radInject.0.Text")).Selected = true;
            }
            else
            {
                radInject.Items.FindByText(GetLocalizedString("radInject.1.Text")).Selected = true;
            }

            chkEnabled.Checked = injection.IsEnabled;
            hidInjectionId.Value = injection.InjectionId.ToString();
            cvName.Enabled = false;

            cmdDelete.Visible = !string.IsNullOrEmpty(hidInjectionId.Value);

            if (injection.CustomProperties.Any(p => p.Name == InjectionInfoMembers.CrmPriorityField))
            {
                var priorityLevel = injection.CustomProperties.FirstOrDefault(p => p.Name == InjectionInfoMembers.CrmPriorityField).Value;
                txtCrmPriority.Text = (priorityLevel == "-1") ? string.Empty : priorityLevel;
            } 
            
            if (injection.CustomProperties.Any(p => p.Name == InjectionInfoMembers.CrmProviderField))
            {
                var provider = injection.CustomProperties.FirstOrDefault(p => p.Name == InjectionInfoMembers.CrmProviderField).Value;
                ddlCrmProvider.ClearSelection();
                ddlCrmProvider.SelectedIndex = int.Parse(provider);
            }

            if (injection.CustomProperties.Any(p => p.Name == InjectionInfoMembers.LastUpdatedByField))
            {
                var updatedBy = injection.CustomProperties.FirstOrDefault(p => p.Name == InjectionInfoMembers.LastUpdatedByField);
                var updatedDate = injection.CustomProperties.FirstOrDefault(p => p.Name == InjectionInfoMembers.LastUpdatedDateField);

                var user = UserController.GetUserById(PortalSettings.PortalId, int.Parse(updatedBy.Value, NumberStyles.Integer));

                lblAudit.Text = string.Format(GetLocalizedString("lblAudit"), user.DisplayName, updatedDate.Value);
            }

            if (injection.CustomProperties.Any(p => p.Name == InjectionInfoMembers.InjectionTypeField))
            {
                var intType =int.Parse(
                    injection.CustomProperties.FirstOrDefault(p => p.Name == InjectionInfoMembers.InjectionTypeField)
                        .Value);

                radType.SelectedIndex = intType;
            }
            else
            {
                // the default for existing injections is 1 because JS/CSS wasn't an option in the past
                radType.SelectedIndex = 1;
            }

            ToggleType();
        }

        private void HandleException(Exception exc)
        {
            Exceptions.LogException(exc);
            if (UserInfo.IsSuperUser | UserInfo.UserID == PortalSettings.AdministratorId)
            {
                DotNetNuke.UI.Skins.Skin.AddModuleMessage(this, string.Concat(exc.Message, "<br />", exc.StackTrace), DotNetNuke.UI.Skins.Controls.ModuleMessage.ModuleMessageType.RedError);
            }
            else
            {
                DotNetNuke.UI.Skins.Skin.AddModuleMessage(this, exc.Message, DotNetNuke.UI.Skins.Controls.ModuleMessage.ModuleMessageType.RedError);
            }
        }

        private void SwapOrder(int ItemId, string UpDown)
        {
            // set the global id to match the one we're looking for
            p_SearchParam = ItemId;

            // change the order
            InjectionController ctlModule = new InjectionController();
            ctlModule.ChangeOrder(ItemId, UpDown);
        }

        #endregion

        #region Data Access

        private void SaveInjection()
        {
            try
            {
                var ctlModule = new InjectionController();
                InjectionInfo objInj = null;

                objInj = !string.IsNullOrEmpty(hidInjectionId.Value) ? ctlModule.GetInjectionContent(int.Parse(hidInjectionId.Value)) : new InjectionInfo();

                PopulateInjectionForSave(ref objInj);

                if (!string.IsNullOrEmpty(hidInjectionId.Value))
                {
                    ctlModule.UpdateInjectionContent(objInj);
                }
                else
                {
                    ctlModule.AddInjectionContent(objInj);
                }
            }
            catch (Exception ex)
            {
                HandleException(ex);
            }
        }

        private void PopulateInjectionForSave(ref InjectionInfo injection)
        {
            var security = new PortalSecurity();

            injection.InjectContent = Server.HtmlEncode(txtContent.Text);
            injection.InjectName = security.InputFilter(txtName.Text, PortalSecurity.FilterFlag.NoMarkup);
            injection.InjectTop = radInject.Items.FindByText(GetLocalizedString("radInject.0.Text")).Selected;
            injection.IsEnabled = chkEnabled.Checked;
            injection.ModuleId = ModuleId;

            if (injection.CustomProperties.Any(p => p.Name == InjectionInfoMembers.InjectionTypeField))
            {
                // update the existing injection type
                injection.CustomProperties.FirstOrDefault(p => p.Name == InjectionInfoMembers.InjectionTypeField)
                    .Value = radType.SelectedIndex.ToString();
            }
            else
            {
                // create new injection type
                injection.CustomProperties.Add(new CustomPropertyInfo()
                {
                    Name = InjectionInfoMembers.InjectionTypeField,
                    Value = radType.SelectedIndex.ToString()
                });
            }

            // CRM PRIORITY LEVEL
            var priorityLevel = ParsePriotityLevel(security);
            if (injection.CustomProperties.Any(p => p.Name == InjectionInfoMembers.CrmPriorityField))
            {
                // update existing CRM/CDF priority level
                injection.CustomProperties.FirstOrDefault(p => p.Name == InjectionInfoMembers.CrmPriorityField)
                    .Value = priorityLevel.ToString();
            }
            else
            {
                // create new CRM/CDF priority level
                injection.CustomProperties.Add(new CustomPropertyInfo()
                {
                    Name = InjectionInfoMembers.CrmPriorityField,
                    Value = priorityLevel.ToString()
                });
            }

            // CRM PAGE PROVIDER
            var provider = ddlCrmProvider.SelectedIndex;
            if (injection.CustomProperties.Any(p => p.Name == InjectionInfoMembers.CrmProviderField))
            {
                // update existing CRM/CDF provider
                injection.CustomProperties.FirstOrDefault(p => p.Name == InjectionInfoMembers.CrmProviderField).Value = provider.ToString();
            }
            else
            {
                // create new CRM/CDF provider
                injection.CustomProperties.Add(new CustomPropertyInfo()
                {
                    Name = InjectionInfoMembers.CrmProviderField,
                    Value = provider.ToString()
                });
            }

            if (injection.CustomProperties.Any(p => p.Name == InjectionInfoMembers.CrmPriorityField))
            {
                injection.CustomProperties.FirstOrDefault(p => p.Name == InjectionInfoMembers.CrmPriorityField)
                    .Value = security.InputFilter(txtCrmPriority.Text.Trim(), PortalSecurity.FilterFlag.NoMarkup);
            }
            else
            {
                injection.CustomProperties.Add(new CustomPropertyInfo()
                {
                    Name = InjectionInfoMembers.CrmPriorityField,
                    Value = security.InputFilter(txtCrmPriority.Text.Trim(), PortalSecurity.FilterFlag.NoMarkup)
                });
            }

            if (injection.CustomProperties.Any(p => p.Name == InjectionInfoMembers.LastUpdatedByField))
            {
                // update the existing auditing fields
                injection.CustomProperties.FirstOrDefault(p => p.Name == InjectionInfoMembers.LastUpdatedByField)
                    .Value = UserInfo.UserID.ToString();
                injection.CustomProperties.FirstOrDefault(p => p.Name == InjectionInfoMembers.LastUpdatedDateField)
                    .Value = DateTime.UtcNow.ToString();
            }
            else
            {
                // adding new audting fields
                injection.CustomProperties.Add(new CustomPropertyInfo()
                {
                    Name = InjectionInfoMembers.LastUpdatedByField,
                    Value = UserInfo.UserID.ToString()
                });

                injection.CustomProperties.Add(new CustomPropertyInfo()
                {
                    Name = InjectionInfoMembers.LastUpdatedDateField,
                    Value = DateTime.UtcNow.ToString()
                });
            }
        }

        private int ParsePriotityLevel(PortalSecurity security)
        {
            var priorityInput = security.InputFilter(txtCrmPriority.Text.Trim(), PortalSecurity.FilterFlag.NoMarkup);
            var priorityLevel = InjectionController.GetCrmPriority(priorityInput);

            return (priorityLevel > Null.NullInteger) ? priorityLevel : Null.NullInteger;
        }

        #endregion

        #region User Interface

        private void ToggleType()
        {
            divInject.Visible = radType.SelectedIndex == 1;
            radInject.Enabled = radType.SelectedIndex == 1;
            cvContent.Enabled = radType.SelectedIndex == 0;
            divAdvanced.Visible = radType.SelectedIndex == 0;

            txtContent.Rows = radType.SelectedIndex == 1 ? 10 : 2;
        }

        private void TogglePanels()
        {
            pnlAddNew.Visible = (!pnlAddNew.Visible);
            pnlManage.Visible = (!pnlManage.Visible);
        }

        protected bool CommandUpVisible(object InjectionId)
        {
            if (InjectionId == null) return false;

            var ctlModule = new InjectionController();
            var oInject = ctlModule.GetInjectionContent(int.Parse(InjectionId.ToString(), NumberStyles.Integer));

            return (oInject.OrderShown != 1);
        }

        protected bool CommandDownVisible(object InjectionId)
        {
            if (InjectionId == null) return false;

            var ctlModule = new InjectionController();
            var oInject = ctlModule.GetInjectionContent(int.Parse(InjectionId.ToString(), NumberStyles.Integer));
            var collInject = ctlModule.GetInjectionContents(ModuleId);

            return (oInject.OrderShown != collInject.Count);
        }

        protected string GetEnabledImage(object EnabledText)
        {
            if (EnabledText != null && string.Equals(EnabledText.ToString(), c_True))
            {
                return EnabledImage;
            }
            else
            {
                return DisabledImage;
            }
        }

        protected string GetEnabledImageAltText(object EnabledText)
        {
            if (EnabledText != null && string.Equals(EnabledText, c_True))
            {
                return EnabledAltText;
            }
            else
            {
                return DisabledAltText;
            }
        }

        protected string GetInjectionTypeForDisplay(object InjectionId)
        {
            if (InjectionId == null) return string.Empty;

            var ctl = new InjectionController();
            var injection = ctl.GetInjectionContent(int.Parse(InjectionId.ToString(), NumberStyles.Integer));
            var injectionType = InjectionController.GetInjectionType(injection);
            
            if (injectionType == InjectionType.HtmlBottom || injectionType == InjectionType.HtmlTop)
            {
                return GetLocalizedString(injectionType.ToString());
            }
            else
            {
                var injectionProvider = InjectionController.GetCrmProvider(injection);

                if (string.IsNullOrEmpty(injectionProvider))
                {
                    return string.Concat(GetLocalizedString(injectionType.ToString()),
                        GetLocalizedString(InjectionController.GetCrmProviderDefault(injectionType)));
                }
                else
                {
                    return string.Concat(GetLocalizedString(injectionType.ToString()),
                        GetLocalizedString(injectionProvider));
                }
            }
        }

        #endregion
    }
}